sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ng/AstBuilder.scala

```diff
@@ -0,0 +1,1128 @@
+        global = false,
+        RepartitionByExpression(expressions, query))
+    } else if (order.isEmpty && sort.isEmpty && distributeBy.isEmpty && clusterBy.isEmpty) {
+      // [EMPTY]
+      query
+    } else {
+      notSupported("Combination of ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY is not supported", ctx)
+    }
+
+    // LIMIT
+    val withLimit = withOrder.optional(limit) {
+      Limit(typedVisit(limit), withOrder)
+    }
+
+    // WINDOWS
+    withLimit.optionalMap(windows)(withWindows)
+  }
+
+  /**
+   * Create a logical plan using a query specification.
+   */
+  override def visitQuerySpecification(
+      ctx: QuerySpecificationContext): LogicalPlan = withOrigin(ctx) {
+    val from = OneRowRelation.optional(ctx.fromClause) {
+      visitFromClause(ctx.fromClause)
+    }
+    withQuerySpecification(ctx, from)
+  }
+
+  /**
+   * Add a query specification to a logical plan. The query specification is the core of the logical
+   * plan, this is where sourcing (FROM clause), transforming (SELECT TRANSFORM/MAP/REDUCE),
+   * projection (SELECT), aggregation (GROUP BY ... HAVING ...) and filtering (WHERE) takes place.
+   *
+   * Note that query hints are ignored (both by the parser and the builder).
+   */
+  private def withQuerySpecification(
+      ctx: QuerySpecificationContext,
+      relation: LogicalPlan): LogicalPlan = withOrigin(ctx) {
+    import ctx._
+
+    // WHERE
+    val withFilter = relation.optional(where) {
+      Filter(expression(where), relation)
+    }
+
+    // Expressions.
+    val expressions = namedExpression.asScala.map(visit).map {
+      case e: Expression => UnresolvedAlias(e)
+    }
+
+    // Create either a transform or a regular query.
+    kind.getType match {
+      case SqlBaseParser.MAP | SqlBaseParser.REDUCE | SqlBaseParser.TRANSFORM =>
+        // Transform
+
+        // Create the attributes.
+        val attributes = if (colTypeList != null) {
+          // Typed return columns.
+          visitColTypeList(colTypeList).toAttributes
+        } else if (columnAliasList != null) {
+          // Untyped return columns.
+          visitColumnAliasList(columnAliasList).map { name =>
+            AttributeReference(name, StringType, nullable = true)()
+          }
+        } else {
+          Seq.empty
+        }
+
+        // Create the transform.
+        ScriptTransformation(
+          expressions,
+          unquote(script.getText),
+          attributes,
+          withFilter,
+          withScriptIOSchema(inRowFormat, outRowFormat, outRecordReader))
+
+      case SqlBaseParser.SELECT =>
+        // Regular select
+
+        // Add lateral views.
+        val withLateralView = ctx.lateralView.asScala.foldLeft(withFilter)(withGenerate)
+
+        // Add aggregation with having or a project.
+        val withProject = if (aggregation != null) {
+          withAggregation(aggregation, expressions, withLateralView).optionalMap(having) {
+            case (h, p) => Filter(expression(h), p)
```

Comment:
Note to self: Group By and Having clause should not be pipelined, a user is allowed to define a global aggregate and filter that...